runtime/drive_handler.go

```diff
@@ -0,0 +1,192 @@
+	ErrVMNotFound = errors.New("vm not found")
+)
+
+// stubDriveHandler is used to create a pool of stubbed file. This will allow
+// different VMs to use the same files and if a VM requires more files than
+// provided, this will handle increasing the files.
+type stubDriveHandler struct {
+	RootPath string
+
+	drives sync.Map
+}
+
+type driveManager struct {
+	stubDriveIndex int64
+	drives         []models.Drive
+}
+
+func newStubDriveHandler(path string) stubDriveHandler {
+	return stubDriveHandler{
+		RootPath: path,
+	}
+}
+
+// StubDrivePaths will create stub drives and return the paths associated with
+// the stub drives.
+func (h *stubDriveHandler) StubDrivePaths(amount int) ([]string, error) {
+	paths := []string{}
+	for i := 0; i < amount; i++ {
+		driveID := fmt.Sprintf("stub%d", i)
+		path := filepath.Join(h.RootPath, driveID)
+
+		f, err := os.OpenFile(path, os.O_CREATE|os.O_EXCL|os.O_RDWR, 0600)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := writeStubID(f, driveID); err != nil {
+			f.Close()
+			return nil, err
+		}
+
+		// Firecracker will not show any drives smaller than 512 bytes. If the
+		// drive size is not a multiple of 512, then the remaining bytes will not
+		// be visible to firecracker.
+		if err := os.Truncate(path, minFCDriveImageBytes); err != nil {
+			f.Close()
+			return nil, err
+		}
+
+		paths = append(paths, path)
+		f.Close()
+	}
+
+	return paths, nil
+}
+
+func writeStubID(w io.Writer, id string) error {
+	length := len(id)
+	if length > 0xFF {
+		return fmt.Errorf("Length of drive id, %d, is too long and is limited to 255 characters", length)
+	}
+
+	if _, err := w.Write([]byte(fmt.Sprintf("%s%c%s", internal.MagicStubBytes, byte(length), id))); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SetDrives will set the given drives based off the VM id. The index
+// represents where the stub drive index starts.
+func (h *stubDriveHandler) SetDrives(id string, index int64, d []models.Drive) {
+	h.drives.Store(id, driveManager{
+		stubDriveIndex: index,
+		drives:         d,
+	})
+}
+
+// GetDrives will return a series of drives related to the given key, which in
+// this case is the VM ID.
+func (h *stubDriveHandler) GetDrives(id string) ([]models.Drive, bool) {
+	drivesIface, ok := h.drives.Load(id)
+	if !ok {
+		return nil, false
+	}
+
+	return drivesIface.(driveManager).drives, true
+}
+
+// InDriveSet will attempt to load the drives associated with the given VM ID.
+// If no VM ID is associated with anything, then false will be returned.  Also,
+// if the drive path is the same as any of the paths in the drives, then true
+// will be returned.
+func (h *stubDriveHandler) InDriveSet(id, path string) bool {
+	drives, ok := h.GetDrives(id)
+	if !ok {
+		return false
+	}
+
+	for _, d := range drives {
+		if firecracker.StringValue(d.PathOnHost) == path {
+			return true
+		}
+	}
+
+	return false
+}
+
+// PatchStubDrive will patch a given drive with the next available stub drive.
```

Comment:
How about: "PatchStubDrive will replace the next available stub drive with the provided drive"?